src/dnnl/dnnl_graph_compiler.cpp

```diff
@@ -1,9 +1,21 @@
 #include "dnnl_graph_compiler.hpp"
+#include "dnnl_graph_compiler_version.h"
 #include <new>
 
 // dnnl_graph_compiler.h interface implementation.
 // TODO: Implement
 
+dnnl_status_t
+dnnl_graph_compiler_get_api_version(dnnl_graph_compiler_api_version *v) {
+  if (!v)
+    return dnnl_invalid_arguments;
+  v->major = GC_API_VERSION_MAJOR;
```

Comment:
Since this is a oneDNN API, I propose using DNNL in the names, e.g.: DNNL_GC_VERSION.

Reply:
One thing that came to my mind: this function is supposed to return the version of the API that it implements, not the version of the graph compiler itself. If there are two versions of the graph compiler (0.2.0 - supports compiling pattern X, 0.1.0 - doesn't support pattern X) there's no way we can know which version we're dealing with since they would likely implement the same version of the API.

There are several things we can do:
1. Return both versions: the version of the API and the version of GC
2. Only return the version of GC and deduce on oneDNN's side which API version was used

With the first approach, it's easier to verify that the loaded library satisfies API requirements (since the current API version is defined at oneDNN's side and we won't have to deduce which API version was used to compile GC's `.so`), however, it will make the version structure more complicated.

cc @kurapov-peter

Reply:
Sure, it should return the version of the API that it implements, not the GC version. That's why I propose having DNNL in the names. The version is defined in the DNNL header and the graph compiler returns the version it implements and was compiled with.

Reply:
> Sure, it should return the version of the API that it implements, not the GC version. That's why I propose having DNNL in the names. The version is defined in the DNNL header and the graph compiler returns the version it implements and was compiled with.

Sounds good to me

Reply:
> Sure, it should return the version of the API that it implements, not the GC version. That's why I propose having DNNL in the names. The version is defined in the DNNL header and the graph compiler returns the version it implements and was compiled with.

Yes, I agree with that. But what are we gonna do with the GC version itself. Is everyone okay with the following schema?

```cpp
// oneDNN: dnnl_graph_compiler.h

#define DNNL_GC_API_MAJOR_V 0
#define DNNL_GC_API_MINOR_V 1
#define DNNL_GC_API_PATCH_V 0
#define DNNL_GC_API_HASH "some_hash"

struct dnnl_graph_compiler_api_version {
    int major, minor, patch;
    const char *hash;
};

struct dnnl_graph_compiler_version {
    int major, minor, patch;
    const char *hash;
    // other meta...
    // int is_gpu_build;
    // int is_cpu_build;
    // ...
};

dnnl_status_t dnnl_graph_compiler_get_api_version(dnnl_graph_compiler_api_version *v);

dnnl_status_t dnnl_graph_compiler_get_version(dnnl_graph_compiler_version *v);

// graph compiler : dnnl_graph_compiler.cpp

dnnl_status_t dnnl_graph_compiler_get_api_version(dnnl_graph_compiler_api_version *v) {
    // fill version with the defines from 'dnnl_graph_compiler.h'
    v->major = DNNL_GC_API_MAJOR_V;
    v->minor = DNNL_GC_API_MINOR_V;
    v->patch = DNNL_GC_API_PATCH_V;
    v->hash = DNNL_GC_API_HASH;
}

dnnl_status_t dnnl_graph_compiler_get_version(dnnl_graph_compiler_version *v) {
    // version of the GC itself, filling with our own numbers
    v->major = 1;
    v->minor = 2;
    v->patch = 3;
    v->hash = "some";
    v->is_gpu_build = 0;
    v->is_cpu_build = 1;
    // ...
}
```

Reply:
What about something like this?

``` c++
struct dnnl_graph_compiler_info {
    int version;
    // Or:
    // int version_major, version_minor, version_patch; 
    const char *git_hash;
    // other meta...
    // int is_gpu_build;
    // int is_cpu_build;
    // ...
};

dnnl_status_t dnnl_graph_compiler_get_info(dnnl_graph_compiler_info *v);
```

Reply:
should `dnnl_graph_compiler_info` also contain info regarding the API version being used or it's only for GC version?

Reply:
It should definitely contain the API version. If the GC library version is required, it could also be added to this structure. If the lib version is not currently required, it could be added in a later version, if there will be such a demand.

Reply:
I think I lost track of the thread. So what we want to achieve is to have the compiler implementation report the version of the graph_compiler.so when queried by the backend through some `get_version()` API. The backend would then check if the version returned is in range of supported. Can we start with something as simple as reporting the compiler version?

Reply:
There is a confusion about 2 versions:
1. The version of the dnnl gc api (dnnl_graph_compiler.h interface), implemented by the graph compiler.
2. The graph compiler version.

From the backend perspective, the first one matters, because this is the API it uses for the GC interaction.
If the second one is not currently required for the backend, I'd not even add it to the structure, because it will be just a dead code.

Reply:
Why do we need two? Can we have just one version and check for it?

Reply:
Assume the following development process:
1. A new function is added to the oneDNN GC API interface. On the backend side a new code, that uses this new function, is implemented. This new code is wrapped into the API version check. At this point it knows nothing about the minimum version of GC, that implements this new function, just because it does not exist yet.
2. On the GC side the new function is implemented and a new version of the library is built with the new version of the oneDNN GC API interface.

Reply:
Implemented approach with two versions in this PR (https://github.com/intel-ai/oneDNN/pull/5)